### DISCUSSION TITLE
Why isn't Ocelot reviewing my Token claims?

### DISCUSSION THREAD

**Discussion by kopticx** [ThreeMammals/Ocelot#1831] [Q&A]
Hello, I am trying to add authentication to my endpoints, however it appears that ocelot is not verifying the claims because even though my token does not have the admin role it still allows me to POST.

How could I resolve this?

![image](https://github.com/ThreeMammals/Ocelot/assets/76010398/1ce0e73c-7ca9-4f2b-8553-acd08dca81dd)

Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJodHRwOi8vc2NoZW1hcy54bWxzb2FwLm9yZy93cy8yMDA1LzA1L2lkZW50aXR5L2NsYWltcy9lbWFpbGFkZHJlc3MiOiJrZXZpbjJAYWEuY29tIiwiaHR0cDovL3NjaGVtYXMueG1sc29hcC5vcmcvd3MvMjAwNS8wNS9pZGVudGl0eS9jbGFpbXMvbmFtZSI6ImtvcHRpY3gyIiwiZXhwIjoxNzMzMzIyMDQzfQ.H1WYrYEuH6EppNnXI2TJASDl1PeO6_943sPIf4FJl0o

![image](https://github.com/ThreeMammals/Ocelot/assets/76010398/d7d8b5c5-efab-4974-b2d6-2ad7064d404d)

![image](https://github.com/ThreeMammals/Ocelot/assets/76010398/260fe646-ac08-4249-9b05-bde18fadfe93)

**Comment by raman-m** [comment]
Hi @kopticx !
>  it appears that **ocelot is not verifying the claims** because even though my token does not have the admin role it still allows me to POST.

Are you kidding?
Have you read the docs: [Authentication](https://ocelot.readthedocs.io/en/latest/features/authentication.html) ?
Have you configured your Ocelot web app to process authentication including Bearer one?
Do you add `AddJwtBearer` call and setup? 😉 

Remove "RouteClaimsRequirement" options from the route JSON and setup Authentication first.

----
>  it still allows me to POST.

It allowed to Post because you have not setup Authentication correctly...
If you develop `catalog.api` downstream service, I recommend you to move all auth logic there and keep route as anonymous one.

**Reply by kopticx** [reply]
I had already configured Bearer authentication
![image](https://github.com/ThreeMammals/Ocelot/assets/76010398/d153fb0d-cdba-49b8-bbd3-43ea5ec76ecd)


This works correctly for my identity Service and for the Get endpoints, but for the POST it is not reading my claim, as I told you, although my claim does not have the admin role, it still allows you to do the POST.

If it were not configured correctly, it would not work with other endpoints, would it?

So What am I supposed to do now?

**Reply by raman-m** [reply]
> So What am I supposed to do now?

Learn ASP.NET Core 8 😉 
- [Overview of ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/introduction-to-aspnet-core?view=aspnetcore-8.0) 
- [Overview of ASP.NET Core authentication](https://learn.microsoft.com/en-us/aspnet/core/security/authentication/?view=aspnetcore-8.0)

Seriously...
I recommend you to play with claims, tokens, roles etc. in another ASP.NET web app solution to design your prototype of authentication with current identity provider.
I have no idea how your identity provider `http://localhost:80005` signs tokens!
Design and check your Auth solution without Ocelot first!
After that you can start implement the same using Ocelot environment...

If you will find real bug in Ocelot Authentication or Authorization features, let us know please.
Good luck in your researches!

**Reply by kopticx** [reply]
The problem with their library is that they continue to do the claims mapping even though it is configured in the project, I had to change the name of my roles for it to work.
